### PR TITLE
Add ProcessFailureError to live_devices with failed process IDs

### PR DESCRIPTION
**Pull request by ymlei**
## Summary

When `live_devices` detects that processes have died during a `with` block, it now raises `ProcessFailureError` instead of a generic exception, so callers can identify exactly which devices failed:

```python
from jax.experimental.multihost_utils import ProcessFailureError

try:
    with live_devices(jax.devices()) as devices:
        ...
except ProcessFailureError as e:
    log(f"Devices {e.failed_devices} failed")
    recover(e.failed_devices)
```

## Changes

- Added `ProcessFailureError(Exception)` class to `jax/experimental/multihost_utils.py` with a `failed_devices: set[Device]` attribute
- In `_LiveDevices.__call__`'s `finally` block, compute failed process IDs from the diff of `old_devices` vs `new_devices` and raise `ProcessFailureError` (chaining any in-block exception via `from exception`)
- Updated `_LiveDevices` docstring `Raises` section

I leave fault_tolerance.rst untouched but would be happy to contribute if needed.

**Edit:** Changed the exception attribute from `failed_process_ids: set[int]` to `failed_devices: set[Device]` per code owner's suggestion.

Enhancement: #36372

### CODE REVIEW

File: jax/experimental/multihost_utils.py [hunk 1]
```diff
@@ -668,6 +682,11 @@ def __call__(self, devices):
       old_devices = self.devices
       new_devices = _live_devices(client, devices)
       self.devices = new_devices
+      failed_process_ids = {
+          d.process_index for d in old_devices if d not in new_devices
+      }
+      if failed_process_ids:
+        raise ProcessFailureError(failed_process_ids) from exception
```
[the remark this answers is not in the document]
Comment (ymlei) [reply jax/experimental/multihost_utils.py]: Without this PR, this branch also seems masked by "if exception:" before it, as process failures mostly always raise exceptions. Unless I overlook something, maybe remove it as it has always been dead code? @mwhittaker

[the remark this answers is not in the document]
Comment (mwhittaker) [reply jax/experimental/multihost_utils.py]: It's possible for a failure to happen but no exception is raised. This documentation describes this in more detail: https://docs.jax.dev/en/latest/fault_tolerance.html#id3. I agree with Gemini that it's dead code now and should be removed.

File: jax/experimental/multihost_utils.py [hunk 2]
```diff
@@ -518,6 +518,17 @@ def _gtl_lowering(ctx, x, *, global_mesh, pspec):
...
 
+class ProcessFailureError(Exception):
+  """Raised by live_devices when one or more processes have failed.
+
+  Attributes:
+    failed_devices: A set of devices whose processes failed.
+  """
+  def __init__(self, failed_devices: set[xla_client.Device]):
+    self.failed_devices = failed_devices
+    super().__init__(f'Processes failed: {failed_devices}')
```
Comment (mwhittaker) [inline jax/experimental/multihost_utils.py]:
```suggestion
    super().__init__(f'Devices failed: {failed_devices}')
```

File: jax/experimental/multihost_utils.py [hunk 3]
```diff
@@ -668,6 +682,9 @@ def __call__(self, devices):
       old_devices = self.devices
       new_devices = _live_devices(client, devices)
       self.devices = new_devices
+      failed_devices = {d for d in old_devices if d not in new_devices}
```
Comment (mwhittaker) [inline jax/experimental/multihost_utils.py]: I think this is not quite right. To understand why, let me explain incarnation ids. Whenever a JAX process starts, it gets assigned a unique randomly generated id called an *incarnation id*. `old_devices` and `new_devices` are dictionaries that map a device to its incarnation. You can read more about this here: https://docs.jax.dev/en/latest/fault_tolerance.html#id3.

It's possible that during a `with live_devices() as devices:` block, a device fails and restarts with a new incarnation. In this case, the device is in `old_devices` and `new_devices`, but they have different incarnations. So the set of failed devices should be the devices that are in `old_devices` but not in `new_devices` or are in `old_devices` but have a different incarnation id in `new_devices`.

File: jax/experimental/multihost_utils.py [hunk 4]
```diff
@@ -668,6 +682,9 @@ def __call__(self, devices):
...
+      if failed_devices:
+        raise ProcessFailureError(failed_devices) from exception
```
Comment (mwhittaker) [inline jax/experimental/multihost_utils.py:692]: Nice! I didn't know about `raise... from...`.

File: jax/experimental/multihost_utils.py (hunk 3 above)
Comment (ymlei) [reply jax/experimental/multihost_utils.py]: Nice catch! Suggested change is updated.

**Review by mwhittaker — APPROVED** [review]

**Comment by ymlei** [comment]
Hi @mwhittaker, quick follow-up — is this blocked on internal checks, or just waiting to be merged? Happy to help if anything's needed on my end. Thanks!

**Comment by mwhittaker** [comment]
@ymlei Oops, sorry about that. Would you mind syncing to HEAD? And then I'll shepherd it more closely and get it merged.

**Comment by ymlei** [comment]
@mwhittaker No worries! Just rebased onto HEAD — branch should now be up to date with main. Thanks for shepherding it through.
